### PR TITLE
Release 5.4 real time data visualization

### PR DESCRIPTION
**Pull request by fstarlike**
### Describe your changes

This pull request introduces a comprehensive real-time data visualization system for WP-SlimStat, along with a modern report architecture that provides a foundation for future report development. The new Live Analytics dashboard provides instant insights into website traffic with auto-refreshing charts and metrics.

**Key Features:**
- Real-time analytics dashboard with live data updates
- Interactive charts showing pageviews, unique visitors, and bounce rates
- Auto-refresh functionality with configurable intervals
- Modern, responsive UI with smooth animations
- Multiple data visualization components (metrics cards, charts, tables)
- Dedicated CSS and JavaScript assets for optimal performance
- New `LiveAnalyticsReport` class implementing the report architecture
- Integration with existing SlimStat report system

**New Report Architecture:**
This PR introduces a modern, OOP-based report registration and management system that provides:
- **Abstract Base Classes**: `AbstractReport`, `ChartReport`, `TableReport`, `SummaryReport` for different report types
- **Interface System**: `ChartableInterface`, `FilterableInterface`, `PaginatableInterface`, etc. for standardized capabilities
- **ReportRegistry**: Centralized registration system for managing all reports
- **LegacyReportAdapter**: Compatibility layer allowing gradual migration of existing reports
- **Extensibility**: Clean hooks and filters for Pro addons and custom reports
- **Type Safety**: Strongly-typed methods and contracts for better IDE support and error prevention

This architecture allows legacy reports to continue working while providing a clear migration path. All new reports should use this standardized structure for consistency, maintainability, and better code organization.

**Technical Changes:**
- Created new `LiveAnalyticsReport.php` with comprehensive real-time data methods
- Added `live-analytics.js` for client-side interactivity and auto-refresh
- Added `live-analytics.css` for modern, responsive styling
- Updated report registration system to include the new live analytics report
- Enhanced chart rendering capabilities for real-time data
- Modified report abstracts to support live data streaming
- Introduced `Bootstrap.php` for report system initialization
- Added `ReportRegistry.php` for centralized report management
- Created `LegacyReportAdapter.php` for backward compatibility

### Submission Review Guidelines:

- ✅ I have performed a self-review of my code
- ✅ If it is a core feature, I have added thorough tests.
- ✅ Will this be part of a product update? If yes, please write one phrase about this update.
  - **"Real-time data visualization with live analytics dashboard and modern report architecture for instant insights and easier report development."**
- ✅ I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- ✅ Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.
- ✅ My code follows the style guidelines of this project
- ✅ I have updated the change-log in `CHANGELOG.md`.

### Type of change

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

---

> [!NOTE]
> Introduce a real-time Live Analytics dashboard with auto-refreshing metrics/charts and a new OOP report system (interfaces, registry, loader) integrated across admin and charts.
> 
> - **Live Analytics (UI/JS/CSS)**:
>   - New real-time dashboard view `views/reports/live-analytics.php` with metrics (users/pages/countries) and bar chart; auto-refresh via `admin/assets/js/live-analytics.js`; styles in `admin/assets/css/live-analytics.css`.
> - **Report Architecture**:
>   - Add OOP report framework (Abstracts, Contracts, Registry, Loader, Adapter) under `src/Reports/*`.
>   - New `LiveAnalyticsReport` (self-contained AJAX, assets, rendering) registered via `Reports\Bootstrap`.
> - **Admin Integration**:
>   - Initialize Reports system in `admin/index.php`; register AJAX; enqueue `jquery-ui-sortable`; customizer drag/drop in `admin/assets/js/admin.js`.
> - **Charts**:
>   - Enhance chart module (`src/Modules/Chart.php`, `admin/assets/js/slimstat-chart.js`, `views/modules/chart-view.php`) with bar chart support, granularity fixes, tooltips/legend tweaks, and caching improvements.
> - **Styling**:
>   - Extend `admin/assets/css/slimstat.css` with bar chart styles and responsive/UI tweaks.
> - **Docs**:
>   - Update `CHANGELOG.md` with real-time analytics, GDPR salted IPs, and AJAX rate limiting.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 796e691cd9d2f36a9e464c759401753db3ff14c9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210718144333194

### CODE REVIEW

**Comment by parhumm** [comment]
@cursor review

File: admin/assets/js/slimstat-chart.js [hunk 1]
```diff
@@ -339,17 +369,30 @@ document.addEventListener("DOMContentLoaded", function () {
...
+        for (var d = 0; d < preparedDatasets.length; d++) {
+            var ds = preparedDatasets[d];
+            if (chartType === "bar") {
+                ds.backgroundColor = ds.backgroundColor || ds.borderColor + "40";
+                ds.borderRadius = 6;
+                ds.borderSkipped = false;
+                ds.categoryPercentage = 0.8;
+                ds.barPercentage = 0.9;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (fstarlike) [reply admin/assets/js/slimstat-chart.js:392]: Fixed

**Comment by parhumm** [comment]
@Cursor review

**Comment by parhumm** [comment]
@claude Review the PR, make sure that's compatible with the add-ons, consider the backward-comoatibiliy, vulnerability and performance issues.

**Comment by parhumm** [comment]
## PR #103 Review: Real-time Data Visualization & Report Architecture

**Branch:** `release-5.4-real-time-data-visualization` → `master`  
**Changes:** 6,438 additions, 75 deletions across 31 files  
**Status:** ⚠️ **CONFLICTING** - Needs merge conflict resolution

---

## Executive Summary

This is a **major architectural enhancement** introducing:
1. **Live Analytics Dashboard** - Real-time data visualization with auto-refresh
2. **New OOP Report Architecture** - Modern report system with abstracts, interfaces, and registry pattern
3. **Enhanced Chart System** - Bar charts, improved tooltips, and better rendering

**Overall Assessment:** ✅ **APPROVED with Minor Recommendations**

---

## 🔒 Security Analysis

### ✅ **Strengths**

1. **Proper Input Sanitization**
   - All user inputs sanitized: `sanitize_text_field()`, `intval()`, `absint()`
   - Example (LiveAnalyticsReport.php:70,592):
     ```php
     $selected_metric = sanitize_text_field( $_GET['metric'] ?? $_POST['metric'] ?? 'users' );
     $requested_metric = sanitize_text_field( $_POST['metric'] ?? 'users' );
     ```

2. **SQL Injection Protection**
   - All queries use `$wpdb->prepare()` with parameterized queries
   - Example (LiveAnalyticsReport.php:179-201):
     ```php
     $sql = $wpdb->prepare("SELECT COUNT(*) FROM ...", $window_start, $window_start, $window_start);
     ```

3. **XSS Prevention**
   - Proper output escaping: `esc_attr()`, `esc_html()`, `esc_js()`
   - Example (live-analytics.php:44,65,119)

4. **CSRF Protection**
   - AJAX endpoints use `wp_verify_nonce()` (LiveAnalyticsReport.php:572)
   - Nonce validation before processing requests

5. **Rate Limiting**
   - AJAX tracking: 10 requests per 5 seconds per IP (Ajax.php:11-17)
   - Live analytics: 60 requests per minute per user/IP (LiveAnalyticsReport.php:640-662)

6. **Input Validation**
   - Whitelist validation for metrics: only accepts 'users', 'pages', 'countries'
   - Regex validation for report_id: `^[a-zA-Z0-9_]+$` (LiveAnalyticsReport.php:601)

### ⚠️ **Minor Concerns**

1. **Direct $_SERVER Access** (LiveAnalyticsReport.php:642)
   ```php
   $ip_address = $_SERVER['REMOTE_ADDR'] ?? '';
   ```
   **Recommendation:** Consider using `$_SERVER['HTTP_X_FORWARDED_FOR']` with validation for proxy environments

2. **Error Information Disclosure** (LiveAnalyticsReport.php:622,630)
   ```php
   'trace' => WP_DEBUG ? $e->getTraceAsString() : null
   ```
   **Status:** ✅ Acceptable - only exposed when WP_DEBUG is enabled

---

## ⚡ Performance Analysis

### ✅ **Optimizations Implemented**

1. **Caching Strategy**
   - Query-level caching via `Query::allowCaching(true, 1)` (1-second TTL)
   - Transient cache for complex JOINs (60-second TTL)
   - Cache invalidation at minute boundaries (LiveAnalyticsReport.php:292-307)

2. **Optimized Queries**
   - Single query for multiple metrics (LiveAnalyticsReport.php:147-152)
   - Efficient use of `DISTINCT` with `COUNT()`
   - Proper indexing assumptions (dt, dt_out, visit_id)

3. **Frontend Optimizations**
   - Canvas-based charts (hardware accelerated)
   - Debounced updates with `AbortController` (live-analytics.js:27)
   - Progressive data loading

### ⚠️ **Performance Concerns**

1. **Complex JOIN Query** (LiveAnalyticsReport.php:325-361)
   - **Issue:** Nested subquery with GROUP BY on large datasets
   - **Impact:** Could be slow with >100K records per 30-minute window
   - **Recommendation:**
     - Add composite index: `INDEX idx_visit_activity (visit_id, dt, dt_out)`
     - Consider summary tables for high-traffic sites
     - Monitor query execution time

2. **Auto-Refresh Load**
   - **Issue:** 60-second refresh interval for all active users
   - **Impact:** Could generate significant DB load during peak hours
   - **Calculation:** 100 concurrent users = 6,000 queries/hour
   - **Recommendation:**
     - Increase default interval to 90-120 seconds
     - Implement server-sent events (SSE) or WebSockets for true real-time updates
     - Add admin setting to configure refresh interval

3. **N+1 Problem Risk** (chart data generation)
   - **Issue:** 30 iterations for chart data points
   - **Status:** ✅ Mitigated by single query with in-memory processing

### 📊 **Load Testing Recommendations**

```sql
-- Test query performance
EXPLAIN SELECT ... FROM {$wpdb->prefix}slim_stats 
WHERE visit_id > 0 AND dt >= UNIX_TIMESTAMP() - 1800;

-- Check index usage
SHOW INDEX FROM wp_slim_stats;

-- Benchmark with realistic data
-- Recommended: 100K+ records
```

---

## 🔄 Backward Compatibility

### ✅ **Excellent Compatibility**

1. **Legacy Adapter Pattern** (LegacyReportAdapter.php)
   - Wraps old reports in new architecture
   - No breaking changes to existing reports
   - Gradual migration path provided

2. **Hook Preservation**
   - All existing hooks maintained
   - New hooks follow WordPress naming conventions:
     - `slimstat_reports_system_initialized`
     - `slimstat_report_registered`
     - `slimstat_live_analytics_load_assets`

3. **Namespace Strategy**
   - New code uses `SlimStat\Reports\*` namespace
   - Old code remains in global namespace
   - No conflicts detected

4. **Database Schema**
   - No schema changes required
   - Uses existing tables (`wp_slim_stats`, `wp_slim_events`)
   - No migrations needed

### ⚠️ **Potential Issues**

1. **JavaScript Global Variables**
   - New `LiveAnalytics` class in global scope
   - **Risk:** Name collision with other plugins
   - **Recommendation:** Wrap in IIFE or use namespace

2. **CSS Class Names**
   - New classes like `.live-analytics-container`, `.clickable-metric`
   - **Risk:** Low - prefixed with `live-analytics` or `slimstat`
   - **Status:** ✅ Acceptable

---

## 🧩 Add-on Compatibility

### ✅ **Extensibility Features**

1. **Filter Hooks** (19 total identified)
   ```php
   apply_filters('slimstat_live_analytics_load_assets', $should_load, $screen);
   apply_filters('slimstat_live_users_cache_ttl', 60);
   apply_filters('slimstat_reports_di_container', $container);
   ```

2. **Action Hooks**
   ```php
   do_action('slimstat_reports_system_initialized');
   do_action('slimstat_report_registered', $report);
   do_action('slimstat_register_custom_reports');
   ```

3. **Custom Report Registration**
   - Clear API via `ReportRegistry::register()` (Bootstrap.php:224-230)
   - Factory pattern for dependency injection
   - Documentation in code comments (Bootstrap.php:124-137)

### ⚠️ **Add-on Considerations**

1. **Pro Add-on Compatibility**
   - **Check Required:** Test with SlimStat Pro, Export add-on, Custom Reports add-on
   - **Areas of Concern:**
     - Report rendering hooks
     - Chart data filters
     - Custom dashboard widgets

2. **Third-Party Themes**
   - **CSS Isolation:** Live Analytics uses `!important` overrides (live-analytics.css:664-676)
   - **Recommendation:** Add theme compatibility layer

3. **Breaking Change Risk: LOW**
   - LegacyReportAdapter ensures old reports continue working
   - New reports opt-in to new architecture

---

## 📁 Code Quality & Architecture

### ✅ **Strengths**

1. **SOLID Principles**
   - Single Responsibility: Each report class has one purpose
   - Open/Closed: Extendable via interfaces
   - Dependency Inversion: Uses DI container (Bootstrap.php:165-177)

2. **Design Patterns**
   - **Singleton:** ReportRegistry (ReportRegistry.php:78-93)
   - **Factory:** ReportFactory
   - **Adapter:** LegacyReportAdapter
   - **Registry:** ReportRegistry

3. **Type Safety**
   - `declare(strict_types=1)` in all new files
   - Type hints for parameters and return types
   - PHPDoc annotations

4. **Error Handling**
   - Try-catch blocks for AJAX handlers (LiveAnalyticsReport.php:608-632)
   - Graceful degradation on errors
   - Proper HTTP status codes (204, 429, 403)

### ⚠️ **Code Smells**

1. **God Method** (LiveAnalyticsReport.php:281-394)
   - `get_users_chart_data()` is 113 lines
   - **Recommendation:** Extract subquery building logic

2. **Magic Numbers**
   - `30 * 60` (30 minutes) appears 8+ times
   - **Recommendation:** Define constants:
     ```php
     const LIVE_WINDOW_MINUTES = 30;
     const CACHE_TTL_SECONDS = 60;
     ```

3. **Duplicate Code**
   - Chart label generation logic repeated
   - **Recommendation:** Extract to utility method

---

## 🐛 Potential Bugs

### ⚠️ **Issues Found**

1. **Timezone Handling** (LiveAnalyticsReport.php:122)
   ```php
   return current_time( 'timestamp' );
   ```
   **Issue:** `current_time('timestamp')` is deprecated in WP 5.3+  
   **Recommendation:** Use `current_datetime()->getTimestamp()`

2. **Race Condition** (live-analytics.js:103-107)
   ```javascript
   this.activityInterval = setInterval(function () {
       localStorage.setItem(storageKey, Date.now().toString());
   }, 60000);
   ```
   **Issue:** Multiple tabs can create duplicate intervals  
   **Recommendation:** Add tab synchronization via BroadcastChannel API

3. **Memory Leak Risk** (live-analytics.js:68)
   ```javascript
   this.activityInterval = setInterval(...);
   ```
   **Issue:** Interval not cleared on destroy  
   **Recommendation:** Add to cleanup in destroy method

---

## 📊 Testing Recommendations

### ✅ **Required Tests**

1. **Security Tests**
   - [ ] SQL injection attempts on metric parameter
   - [ ] XSS payloads in report_id
   - [ ] CSRF token validation
   - [ ] Rate limiting bypass attempts

2. **Performance Tests**
   - [ ] Load test with 1M+ records
   - [ ] Concurrent user stress test (100+ users)
   - [ ] Query execution time profiling
   - [ ] Memory usage monitoring

3. **Compatibility Tests**
   - [ ] Test with WordPress 5.9, 6.0, 6.1, 6.2+
   - [ ] PHP 7.4, 8.0, 8.1, 8.2 compatibility
   - [ ] MySQL 5.7, 8.0 compatibility
   - [ ] MariaDB 10.3+ compatibility

4. **Add-on Tests**
   - [ ] SlimStat Pro integration
   - [ ] Export add-on compatibility
   - [ ] Custom Reports add-on

---

## ✅ Final Recommendations

### **Critical (Must Fix Before Merge)**

1. **Resolve merge conflicts** (PR shows CONFLICTING status)
2. **Add database indexes:**
   ```sql
   ALTER TABLE wp_slim_stats 
   ADD INDEX idx_visit_activity (visit_id, dt, dt_out);
   ```

### **High Priority (Recommended)**

3. **Increase default refresh interval** from 60s to 90-120s
4. **Fix timezone handling** - Replace `current_time('timestamp')` with `current_datetime()->getTimestamp()`
5. **Add theme compatibility CSS** to prevent conflicts
6. **Test with Pro add-ons** before release

### **Medium Priority (Nice to Have)**

7. **Extract constants** for magic numbers (30 minutes, cache TTL)
8. **Add admin setting** for refresh interval configuration
9. **Implement tab synchronization** for localStorage updates
10. **Add load testing** documentation for high-traffic sites

### **Low Priority (Future Enhancement)**

11. **Consider WebSockets/SSE** for true real-time updates (reduce DB load)
12. **Add summary tables** for performance optimization on large datasets
13. **Implement query result pagination** for chart data

---

## 📝 Verdict

**✅ APPROVED FOR MERGE** (after resolving merge conflicts and adding database index)

### **Scores:**
- **Security:** ⭐⭐⭐⭐⭐ (5/5) - Excellent
- **Performance:** ⭐⭐⭐⭐☆ (4/5) - Good with minor concerns
- **Backward Compatibility:** ⭐⭐⭐⭐⭐ (5/5) - Excellent
- **Code Quality:** ⭐⭐⭐⭐☆ (4/5) - Very Good
- **Add-on Compatibility:** ⭐⭐⭐⭐☆ (4/5) - Good (needs testing)

### **Overall:** ⭐⭐⭐⭐☆ (4.4/5)

This is a **well-architected feature** with strong security practices and good backward compatibility. The main concerns are performance optimization for high-traffic sites and thorough add-on compatibility testing.

---

**Reviewed Files:** 31 modified, focusing on:
- `src/Reports/Types/Analytics/LiveAnalyticsReport.php` (718 lines)
- `src/Reports/Bootstrap.php` (276 lines)
- `admin/assets/js/live-analytics.js` (1361 lines)
- `src/Tracker/Ajax.php` (155 lines)
- `src/Utils/Query.php` (1378 lines)
